CMakeLists.txt

```diff
@@ -44,16 +52,28 @@ macro(build_pybind11)
   endif()
   list(APPEND extra_cmake_args "-DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}")
 
-
   include(ExternalProject)
-  ExternalProject_Add(pybind11-2.5.0
-    URL https://github.com/pybind/pybind11/archive/v2.5.0.tar.gz
-    URL_MD5 1ad2c611378fb440e8550a7eb6b31b89
-    TIMEOUT 600
+  ExternalProject_Add(pybind11-2.7.1
+    GIT_REPOSITORY https://github.com/pybind/pybind11.git
+    GIT_TAG v2.7.1
+    GIT_CONFIG advice.detachedHead=false
+    # Suppress git update due to https://gitlab.kitware.com/cmake/cmake/-/issues/16419
+    # See https://github.com/ament/uncrustify_vendor/pull/22 for details
+    UPDATE_COMMAND ""
+    TIMEOUT 300
     CMAKE_ARGS
       -DCMAKE_INSTALL_PREFIX=${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}_install
       -DPYBIND11_TEST=OFF
       ${extra_cmake_args}
+    # By default, pybind11 sets Py_DEBUG if the interpreter is a debug one (as
+    # it is in Windows Debug).  Unfortunately, this conflicts with Py_DEBUG as
+    # used by the CPython headers, and ultimately ends up causing multiple
+    # definitions for Py_DEBUG, which MSVC complains about.  This patch switches
+    # the internal pybind11 variable to be called PYBIND11_DEBUG, which avoids
+    # the issue.
+    PATCH_COMMAND
```

Comment:
CMake has a warning about patching being non-intuitive when using the `git` download method. I think this would only be an issue if Pybind11 changed what the `v2.7.1` tag pointed at, or if someone changed the git tag locally and rebuilt pybind11_vendor.

PATCH_COMMAND <cmd>...

>    Note that it can be quite difficult to define an appropriate patch command that performs robustly, especially for download methods such as git where changing the GIT_TAG will not discard changes from a previous patch, but the patch command will be called again after updating to the new tag.

Reply:
Yes, this is completely broken as far as I can tell.  Right now, if packages using this method ever try to rebuild, they fail to apply the patch and fail the build.  Honestly I think `ExternalProject_add` isn't fit for the purpose we are using it for here, but again this is a problem throughout all of our vendor packages.  I'd love a way to fix it (which probably involves not using `ExternalProject_add`, but that is a larger change).